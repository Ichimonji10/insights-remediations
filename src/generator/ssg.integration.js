'use strict';

const { request, reqId } = require('../test');

test('generates a simple playbook with single compliance remediation', async () => {
    const data = {
        issues: [{
            id: 'compliance:sshd_disable_root_login',
            systems: ['68799a02-8be9-11e8-9eb6-529269fb1459']
        }]
    };

    const res = await request
    .post('/v1/playbook')
    .send(data)
    .expect(200);
    expect(res.text).toMatchSnapshot();
});

test('generates a simple playbook with multiple compliance remediation', async () => {
    const data = {
        issues: [{
            id: 'compliance:no_empty_passwords',
            systems: ['68799a02-8be9-11e8-9eb6-529269fb1459']
        }, {
            id: 'compliance:bootloader_audit_argument',
            systems: ['68799a02-8be9-11e8-9eb6-529269fb1459']
        }]
    };

    const res = await request
    .post('/v1/playbook')
    .send(data)
    .expect(200);
    expect(res.text).toMatchSnapshot();
});

test('generates a simple playbook with reboot support', async () => {
    const data = {
        issues: [{
            id: 'compliance:security_patches_up_to_date',
            systems: ['68799a02-8be9-11e8-9eb6-529269fb1459']
        }]
    };

    const res = await request
    .post('/v1/playbook')
    .send(data)
    .expect(200);
    expect(res.text).toMatchSnapshot();
});

test('400s on unknown resolution type', () => {
    const {id, header} = reqId();

    return request
    .post('/v1/playbook')
    .set(header)
    .send({
        issues: [{
            id: 'compliance:non-existing-issue',
            systems: ['68799a02-8be9-11e8-9eb6-529269fb1459']
        }]
    })
    .expect(400)
    .then(({ body }) => {
        body.errors.should.eql([{
            id,
            status: 400,
            code: 'UNSUPPORTED_ISSUE',
            title: 'Issue "compliance:non-existing-issue" does not have Ansible support'
        }]);
    });
});

test('400s on unknown resolution type other than fix', () => {
    const {id, header} = reqId();

    return request
    .post('/v1/playbook')
    .set(header)
    .send({
        issues: [{
            id: 'compliance:sshd_disable_root_login',
            systems: ['68799a02-8be9-11e8-9eb6-529269fb1459'],
            resolution: 'non-existing-resolution'
        }]
    })
    .expect(400)
    .then(({ body }) => {
        body.errors.should.eql([{
            id,
            status: 400,
            code: 'UNKNOWN_RESOLUTION',
            title: 'Issue "compliance:sshd_disable_root_login"' +
                ' does not have Ansible resolution "non-existing-resolution"'
        }]);
    });
});

