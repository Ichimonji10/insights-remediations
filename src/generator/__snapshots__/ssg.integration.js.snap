// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates a simple playbook with multiple compliance remediation 1`] = `
"---
# Prevent Log In to Accounts With Empty Password
# Identifier: (compliance:xccdf_org.ssgproject.content_rule_no_empty_passwords,fix)
# Version: unknown
- name: fix xccdf_org.ssgproject.content_rule_no_empty_passwords
  hosts: '68799a02-8be9-11e8-9eb6-529269fb1459.example.com'
  become: true
  tasks:
    - name: Prevent Log In to Accounts With Empty Password - system-auth
      replace:
        dest: /etc/pam.d/system-auth
        follow: 'yes'
        regexp: nullok
    - name: Prevent Log In to Accounts With Empty Password - password-auth
      replace:
        dest: /etc/pam.d/password-auth
        follow: 'yes'
        regexp: nullok

# Enable Auditing for Processes Which Start Prior to the Audit Daemon
# Identifier: (compliance:xccdf_org.ssgproject.content_rule_bootloader_audit_argument,fix)
# Version: unknown
- name: fix xccdf_org.ssgproject.content_rule_bootloader_audit_argument
  hosts: '68799a02-8be9-11e8-9eb6-529269fb1459.example.com'
  become: true
  tasks:
    - name: Enable Auditing for Processes Which Start Prior to the Audit Daemon
      shell: /sbin/grubby --update-kernel=ALL --args=\\"audit=1\\"
      register: result
    - when: result is changed
      set_fact:
        insights_needs_reboot: true

# Reboots a system if any of the preceeding plays sets the 'insights_needs_reboot' variable to true.
# The variable can be overridden to suppress this behavior.
- name: Reboot system (if applicable)
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  become: True
  gather_facts: False
  tasks:
    - when:
        - insights_needs_reboot is defined
        - insights_needs_reboot
      block:
        - name: Reboot system
          shell: sleep 2 && shutdown -r now \\"Ansible triggered reboot\\"
          async: 1
          poll: 0
          ignore_errors: true

        - name: Wait for system to boot up
          local_action:
            module: wait_for
            host: \\"{{ hostvars[inventory_hostname]['ansible_host'] | default(hostvars[inventory_hostname]['ansible_ssh_host'], true) | default(inventory_hostname, true) }}\\"
            port: \\"{{ hostvars[inventory_hostname]['ansible_port'] | default(hostvars[inventory_hostname]['ansible_ssh_port'], true) | default('22', true) }}\\"
            delay: 15
            search_regex: OpenSSH
            timeout: 300
          become: false

- name: run insights
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  become: True
  gather_facts: False
  tasks:
    - name: run insights
      command: redhat-access-insights
      changed_when: false"
`;

exports[`generates a simple playbook with reboot support 1`] = `
"---
# System security patches and updates must be installed and up-to-date.
# Identifier: (compliance:xccdf_org.ssgproject.content_rule_security_patches_up_to_date,fix)
# Version: unknown
- name: fix xccdf_org.ssgproject.content_rule_security_patches_up_to_date
  hosts: '68799a02-8be9-11e8-9eb6-529269fb1459.example.com'
  become: true
  tasks:
    - name: Security patches are up to date
      package:
        name: '*'
        state: latest
      register: result
    - when: result is changed
      set_fact:
        insights_needs_reboot: true

# Reboots a system if any of the preceeding plays sets the 'insights_needs_reboot' variable to true.
# The variable can be overridden to suppress this behavior.
- name: Reboot system (if applicable)
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  become: True
  gather_facts: False
  tasks:
    - when:
        - insights_needs_reboot is defined
        - insights_needs_reboot
      block:
        - name: Reboot system
          shell: sleep 2 && shutdown -r now \\"Ansible triggered reboot\\"
          async: 1
          poll: 0
          ignore_errors: true

        - name: Wait for system to boot up
          local_action:
            module: wait_for
            host: \\"{{ hostvars[inventory_hostname]['ansible_host'] | default(hostvars[inventory_hostname]['ansible_ssh_host'], true) | default(inventory_hostname, true) }}\\"
            port: \\"{{ hostvars[inventory_hostname]['ansible_port'] | default(hostvars[inventory_hostname]['ansible_ssh_port'], true) | default('22', true) }}\\"
            delay: 15
            search_regex: OpenSSH
            timeout: 300
          become: false

- name: run insights
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  become: True
  gather_facts: False
  tasks:
    - name: run insights
      command: redhat-access-insights
      changed_when: false"
`;

exports[`generates a simple playbook with single compliance remediation 1`] = `
"---
# Disable SSH Root Login
# Identifier: (compliance:xccdf_org.ssgproject.content_rule_sshd_disable_root_login,fix)
# Version: unknown
- name: fix xccdf_org.ssgproject.content_rule_sshd_disable_root_login
  hosts: '68799a02-8be9-11e8-9eb6-529269fb1459.example.com'
  become: true
  tasks:
    - name: Disable SSH Root Login
      lineinfile:
        create: 'yes'
        dest: /etc/ssh/sshd_config
        regexp: ^PermitRootLogin
        line: PermitRootLogin no
        insertafter: '(?i)^#?authentication'
        validate: sshd -t -f %s

- name: run insights
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  become: True
  gather_facts: False
  tasks:
    - name: run insights
      command: redhat-access-insights
      changed_when: false"
`;
